{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ensembl gene table -> bionty.Gene().df\n",
    "\n",
    "HGNC table contain only 20k+ validated human entries. When people perform bioinformatic analysis, they often use ensembl ids for transcripts and therefore we are switching to the Ensembl table in `bionty.Gene().df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from lnschema_bionty import id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ensembl download\n",
    "\n",
    "These tables are downloaded from [biomart](https://www.ensembl.org/biomart/martview/) database (`Ensembl Genes 107`) containing the following id columns for every species:\n",
    "- `Gene stable ID`\n",
    "- `Transcript stable ID`\n",
    "- `Protein stable ID`\n",
    "- `Gene name`\n",
    "- `Gene Synonym`\n",
    "- `Gene type`\n",
    "- `NCBI gene (formerly Entrezgene) ID`\n",
    "\n",
    "Addtional species-specific columns are also present for:\n",
    "- human: `HGNC ID`, `MIM gene accession`\n",
    "- mouse: `MGI ID`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = {\n",
    "    \"human\": \"~/Downloads/mart_export-human.txt\",  # http://www.ensembl.org/biomart/martview/4d75e3d44de27e6ed58cfce974f0a755?VIRTUALSCHEMANAME=default&ATTRIBUTES=hsapiens_gene_ensembl.default.feature_page.ensembl_gene_id|hsapiens_gene_ensembl.default.feature_page.ensembl_transcript_id|hsapiens_gene_ensembl.default.feature_page.ensembl_peptide_id|hsapiens_gene_ensembl.default.feature_page.external_gene_name|hsapiens_gene_ensembl.default.feature_page.external_synonym|hsapiens_gene_ensembl.default.feature_page.gene_biotype|hsapiens_gene_ensembl.default.feature_page.entrezgene_id|hsapiens_gene_ensembl.default.feature_page.hgnc_id|hsapiens_gene_ensembl.default.feature_page.mim_gene_accession&FILTERS=&VISIBLEPANEL=resultspanel\n",
    "    \"mouse\": \"~/Downloads/mart_export-mouse.txt\",  # http://www.ensembl.org/biomart/martview/4d75e3d44de27e6ed58cfce974f0a755?VIRTUALSCHEMANAME=default&ATTRIBUTES=mmusculus_gene_ensembl.default.feature_page.ensembl_gene_id|mmusculus_gene_ensembl.default.feature_page.ensembl_transcript_id|mmusculus_gene_ensembl.default.feature_page.ensembl_peptide_id|mmusculus_gene_ensembl.default.feature_page.external_gene_name|mmusculus_gene_ensembl.default.feature_page.external_synonym|mmusculus_gene_ensembl.default.feature_page.gene_biotype|mmusculus_gene_ensembl.default.feature_page.entrezgene_id|mmusculus_gene_ensembl.default.feature_page.mgi_id&FILTERS=&VISIBLEPANEL=resultspanel\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate the tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------human----------\n",
      "Initial shape: (620902, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>MIM gene accession</th>\n",
       "      <th>Gene Synonym</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Wn6jXH</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7481</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MTTF|trnF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>balylO</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7470</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12S|MOTS-c|MTRNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3tmE29</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MTTV|trnV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>jviPwd</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7471</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16S|HN|MTRNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RWi7gE</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MTTL1|TRNL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Transcript stable ID Protein stable ID Gene name  \\\n",
       "id                                                                         \n",
       "Wn6jXH  ENSG00000210049      ENST00000387314               NaN     MT-TF   \n",
       "balylO  ENSG00000211459      ENST00000389680               NaN   MT-RNR1   \n",
       "3tmE29  ENSG00000210077      ENST00000387342               NaN     MT-TV   \n",
       "jviPwd  ENSG00000210082      ENST00000387347               NaN   MT-RNR2   \n",
       "RWi7gE  ENSG00000209082      ENST00000386347               NaN    MT-TL1   \n",
       "\n",
       "       Gene type NCBI gene (formerly Entrezgene) ID    HGNC ID  \\\n",
       "id                                                               \n",
       "Wn6jXH   Mt_tRNA                                NaN  HGNC:7481   \n",
       "balylO   Mt_rRNA                                NaN  HGNC:7470   \n",
       "3tmE29   Mt_tRNA                                NaN  HGNC:7500   \n",
       "jviPwd   Mt_rRNA                                NaN  HGNC:7471   \n",
       "RWi7gE   Mt_tRNA                                NaN  HGNC:7490   \n",
       "\n",
       "       MIM gene accession       Gene Synonym  \n",
       "id                                            \n",
       "Wn6jXH                NaN          MTTF|trnF  \n",
       "balylO                NaN  12S|MOTS-c|MTRNR1  \n",
       "3tmE29                NaN          MTTV|trnV  \n",
       "jviPwd                NaN      16S|HN|MTRNR2  \n",
       "RWi7gE                NaN        MTTL1|TRNL1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final shape: (276652, 9)\n",
      "Saved as ensembl-ids-human.parquet.\n",
      "----------mouse----------\n",
      "Initial shape: (296054, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Protein stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "      <th>MGI ID</th>\n",
       "      <th>Gene Synonym</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EtAapO</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>ENSMUST00000082387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:102487</td>\n",
       "      <td>tRNA|tRNA-Phe|TrnF tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cywSTr</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>ENSMUST00000082388</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:102493</td>\n",
       "      <td>12S ribosomal RNA|12S rRNA|12SrRNA|Rnr1 s-rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B60gmR</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>ENSMUST00000082389</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:102472</td>\n",
       "      <td>tRNA|tRNA-Val|TrnaV tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n0jWMu</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>ENSMUST00000082390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:102492</td>\n",
       "      <td>16S ribosomal RNA|16S rRNA|16SrRNA|Rnr2 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k2kfby</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>ENSMUST00000082391</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MGI:102482</td>\n",
       "      <td>tRNA|tRNA Leu|tRNA Leu_1|TrnrL1 tRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene stable ID Transcript stable ID Protein stable ID Gene name  \\\n",
       "id                                                                            \n",
       "EtAapO  ENSMUSG00000064336   ENSMUST00000082387               NaN     mt-Tf   \n",
       "cywSTr  ENSMUSG00000064337   ENSMUST00000082388               NaN   mt-Rnr1   \n",
       "B60gmR  ENSMUSG00000064338   ENSMUST00000082389               NaN     mt-Tv   \n",
       "n0jWMu  ENSMUSG00000064339   ENSMUST00000082390               NaN   mt-Rnr2   \n",
       "k2kfby  ENSMUSG00000064340   ENSMUST00000082391               NaN    mt-Tl1   \n",
       "\n",
       "       Gene type NCBI gene (formerly Entrezgene) ID      MGI ID  \\\n",
       "id                                                                \n",
       "EtAapO   Mt_tRNA                                NaN  MGI:102487   \n",
       "cywSTr   Mt_rRNA                                NaN  MGI:102493   \n",
       "B60gmR   Mt_tRNA                                NaN  MGI:102472   \n",
       "n0jWMu   Mt_rRNA                                NaN  MGI:102492   \n",
       "k2kfby   Mt_tRNA                                NaN  MGI:102482   \n",
       "\n",
       "                                             Gene Synonym  \n",
       "id                                                         \n",
       "EtAapO                            tRNA|tRNA-Phe|TrnF tRNA  \n",
       "cywSTr     12S ribosomal RNA|12S rRNA|12SrRNA|Rnr1 s-rRNA  \n",
       "B60gmR                           tRNA|tRNA-Val|TrnaV tRNA  \n",
       "n0jWMu  16S ribosomal RNA|16S rRNA|16SrRNA|Rnr2 16S ri...  \n",
       "k2kfby               tRNA|tRNA Leu|tRNA Leu_1|TrnrL1 tRNA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final shape: (150702, 8)\n",
      "Saved as ensembl-ids-mouse.parquet.\n"
     ]
    }
   ],
   "source": [
    "for species, path in dfs.items():\n",
    "    print(f\"----------{species}----------\")\n",
    "    df = pd.read_csv(path, dtype=str)\n",
    "    print(f\"Initial shape: {df.shape}\")\n",
    "\n",
    "    # Aggregate the `Gene Synonym` column\n",
    "    df_alias = df[[\"Gene name\", \"Gene Synonym\"]].drop_duplicates().dropna()\n",
    "    df_alias = df_alias.groupby(\"Gene name\").agg(\"|\".join)\n",
    "    del df[\"Gene Synonym\"]\n",
    "    df = df.drop_duplicates()\n",
    "    df = pd.merge(df, df_alias, on=\"Gene name\", how=\"left\")\n",
    "\n",
    "    # add ids to each entry\n",
    "    ids = []\n",
    "    for i in df.index:\n",
    "        ids.append(id.gene())\n",
    "    df.index = ids\n",
    "    df.index.name = \"id\"\n",
    "\n",
    "    display(df.head())\n",
    "    print(f\"Final shape: {df.shape}\")\n",
    "\n",
    "    # save to a feather file\n",
    "    df.to_parquet(f\"ensembl-ids-{species}.parquet\")\n",
    "    print(f\"Saved as ensembl-ids-{species}.parquet.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('py39')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ae1fefc8646a06dd2e75004cd934adda7c5727b046986a772e3b44b0ffba9754"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
