{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1d7b3b73-75f9-468d-8783-873ebadcab9b",
   "metadata": {},
   "source": [
    "# Ingest complete HGNC, MGI gene set from EBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ec6631-8473-4a2d-b488-def921bb83de",
   "metadata": {},
   "outputs": [],
   "source": [
    "from nbproject import header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af510a69-9892-4346-8d2f-567d1d4b944d",
   "metadata": {},
   "outputs": [],
   "source": [
    "HGNC = \"http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d02069e2-07a8-417c-8b41-92a7dc83423e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff8a5421-a7d1-4d13-85fd-e4e12a0c020f",
   "metadata": {},
   "source": [
    "The following call was originally used to download the file, but it times out very often."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96295cb1-2a7f-4648-bb0b-6f191514a0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    HGNC,\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    low_memory=False,  # If True, gets DtypeWarning\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7217ba4-497f-4d77-8986-45bfc4087b49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43160, 53)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cbf2d52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>prev_symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_ids</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hgnc_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HGNC:5</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>ENST00000263100.8|NM_130786.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:37133</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCRNA00181|A1BGAS|A1BG-AS</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:24086</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>ENST00000373997.8|NM_014576.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:7</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>ENST00000318602.12|NM_000014.6</td>\n",
       "      <td>HGNC:7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:27057</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              symbol                            name          locus_group  \\\n",
       "hgnc_id                                                                     \n",
       "HGNC:5          A1BG          alpha-1-B glycoprotein  protein-coding gene   \n",
       "HGNC:37133  A1BG-AS1            A1BG antisense RNA 1       non-coding RNA   \n",
       "HGNC:24086      A1CF  APOBEC1 complementation factor  protein-coding gene   \n",
       "HGNC:7           A2M           alpha-2-macroglobulin  protein-coding gene   \n",
       "HGNC:27057   A2M-AS1             A2M antisense RNA 1       non-coding RNA   \n",
       "\n",
       "                           locus_type    status  location location_sortable  \\\n",
       "hgnc_id                                                                       \n",
       "HGNC:5      gene with protein product  Approved  19q13.43          19q13.43   \n",
       "HGNC:37133       RNA, long non-coding  Approved  19q13.43          19q13.43   \n",
       "HGNC:24086  gene with protein product  Approved  10q11.23          10q11.23   \n",
       "HGNC:7      gene with protein product  Approved  12p13.31          12p13.31   \n",
       "HGNC:27057       RNA, long non-coding  Approved  12p13.31          12p13.31   \n",
       "\n",
       "                             alias_symbol alias_name  \\\n",
       "hgnc_id                                                \n",
       "HGNC:5                                NaN        NaN   \n",
       "HGNC:37133                       FLJ23569        NaN   \n",
       "HGNC:24086  ACF|ASP|ACF64|ACF65|APOBEC1CF        NaN   \n",
       "HGNC:7               FWP007|S863-7|CPAMD5        NaN   \n",
       "HGNC:27057                            NaN        NaN   \n",
       "\n",
       "                          prev_symbol  ...   cd lncrnadb enzyme_id  \\\n",
       "hgnc_id                                ...                           \n",
       "HGNC:5                            NaN  ...  NaN      NaN       NaN   \n",
       "HGNC:37133  NCRNA00181|A1BGAS|A1BG-AS  ...  NaN      NaN       NaN   \n",
       "HGNC:24086                        NaN  ...  NaN      NaN       NaN   \n",
       "HGNC:7                            NaN  ...  NaN      NaN       NaN   \n",
       "HGNC:27057                        NaN  ...  NaN      NaN       NaN   \n",
       "\n",
       "           intermediate_filament_db rna_central_ids lncipedia gtrnadb  \\\n",
       "hgnc_id                                                                 \n",
       "HGNC:5                          NaN             NaN       NaN     NaN   \n",
       "HGNC:37133                      NaN             NaN  A1BG-AS1     NaN   \n",
       "HGNC:24086                      NaN             NaN       NaN     NaN   \n",
       "HGNC:7                          NaN             NaN       NaN     NaN   \n",
       "HGNC:27057                      NaN             NaN   A2M-AS1     NaN   \n",
       "\n",
       "                   agr                     mane_select   gencc  \n",
       "hgnc_id                                                         \n",
       "HGNC:5          HGNC:5   ENST00000263100.8|NM_130786.4     NaN  \n",
       "HGNC:37133  HGNC:37133                             NaN     NaN  \n",
       "HGNC:24086  HGNC:24086   ENST00000373997.8|NM_014576.4     NaN  \n",
       "HGNC:7          HGNC:7  ENST00000318602.12|NM_000014.6  HGNC:7  \n",
       "HGNC:27057  HGNC:27057                             NaN     NaN  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cc43e789-6d19-4253-b262-8420a92d144d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index().to_feather(\"hgnc_complete_set.feather\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19907a51",
   "metadata": {},
   "source": [
    "## Mouse: MGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "53648c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"http://ftp.ensembl.org/pub/release-106/tsv/homo_sapiens/Homo_sapiens.GRCh38.106.uniprot.tsv.gz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a78c54f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>transcript_stable_id</th>\n",
       "      <th>protein_stable_id</th>\n",
       "      <th>xref</th>\n",
       "      <th>db_name</th>\n",
       "      <th>info_type</th>\n",
       "      <th>source_identity</th>\n",
       "      <th>xref_identity</th>\n",
       "      <th>linkage_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>ENST00000379236</td>\n",
       "      <td>ENSP00000368538</td>\n",
       "      <td>P43489</td>\n",
       "      <td>Uniprot/SWISSPROT</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>ENST00000328596</td>\n",
       "      <td>ENSP00000328207</td>\n",
       "      <td>Q9Y5U5</td>\n",
       "      <td>Uniprot/SWISSPROT</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>ENST00000328596</td>\n",
       "      <td>ENSP00000328207</td>\n",
       "      <td>Q9Y5U5-2</td>\n",
       "      <td>Uniprot_isoform</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>ENST00000379268</td>\n",
       "      <td>ENSP00000368570</td>\n",
       "      <td>Q9Y5U5</td>\n",
       "      <td>Uniprot/SWISSPROT</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>ENST00000379268</td>\n",
       "      <td>ENSP00000368570</td>\n",
       "      <td>Q9Y5U5-1</td>\n",
       "      <td>Uniprot_isoform</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156956</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>ENSP00000354665</td>\n",
       "      <td>P03923</td>\n",
       "      <td>Uniprot/SWISSPROT</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156957</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>ENSP00000354665</td>\n",
       "      <td>A0A1X7RCR6</td>\n",
       "      <td>Uniprot/SPTREMBL</td>\n",
       "      <td>SEQUENCE_MATCH</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156958</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>ENSP00000354665</td>\n",
       "      <td>U5Z977</td>\n",
       "      <td>Uniprot/SPTREMBL</td>\n",
       "      <td>SEQUENCE_MATCH</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156959</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>ENSP00000354554</td>\n",
       "      <td>P00156</td>\n",
       "      <td>Uniprot/SWISSPROT</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156960</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>ENSP00000354554</td>\n",
       "      <td>Q0ZFD6</td>\n",
       "      <td>Uniprot/SPTREMBL</td>\n",
       "      <td>SEQUENCE_MATCH</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156961 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_stable_id transcript_stable_id protein_stable_id        xref  \\\n",
       "0       ENSG00000186827      ENST00000379236   ENSP00000368538      P43489   \n",
       "1       ENSG00000186891      ENST00000328596   ENSP00000328207      Q9Y5U5   \n",
       "2       ENSG00000186891      ENST00000328596   ENSP00000328207    Q9Y5U5-2   \n",
       "3       ENSG00000186891      ENST00000379268   ENSP00000368570      Q9Y5U5   \n",
       "4       ENSG00000186891      ENST00000379268   ENSP00000368570    Q9Y5U5-1   \n",
       "...                 ...                  ...               ...         ...   \n",
       "156956  ENSG00000198695      ENST00000361681   ENSP00000354665      P03923   \n",
       "156957  ENSG00000198695      ENST00000361681   ENSP00000354665  A0A1X7RCR6   \n",
       "156958  ENSG00000198695      ENST00000361681   ENSP00000354665      U5Z977   \n",
       "156959  ENSG00000198727      ENST00000361789   ENSP00000354554      P00156   \n",
       "156960  ENSG00000198727      ENST00000361789   ENSP00000354554      Q0ZFD6   \n",
       "\n",
       "                  db_name       info_type source_identity xref_identity  \\\n",
       "0       Uniprot/SWISSPROT          DIRECT             100           100   \n",
       "1       Uniprot/SWISSPROT          DIRECT               -             -   \n",
       "2         Uniprot_isoform          DIRECT               -             -   \n",
       "3       Uniprot/SWISSPROT          DIRECT             100           100   \n",
       "4         Uniprot_isoform          DIRECT               -             -   \n",
       "...                   ...             ...             ...           ...   \n",
       "156956  Uniprot/SWISSPROT          DIRECT             100           100   \n",
       "156957   Uniprot/SPTREMBL  SEQUENCE_MATCH             100           100   \n",
       "156958   Uniprot/SPTREMBL  SEQUENCE_MATCH             100           100   \n",
       "156959  Uniprot/SWISSPROT          DIRECT             100           100   \n",
       "156960   Uniprot/SPTREMBL  SEQUENCE_MATCH             100           100   \n",
       "\n",
       "       linkage_type  \n",
       "0                 -  \n",
       "1                 -  \n",
       "2                 -  \n",
       "3                 -  \n",
       "4                 -  \n",
       "...             ...  \n",
       "156956            -  \n",
       "156957            -  \n",
       "156958            -  \n",
       "156959            -  \n",
       "156960            -  \n",
       "\n",
       "[156961 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b28768bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"http://ftp.ensembl.org/pub/release-106/tsv/homo_sapiens/Homo_sapiens.GRCh38.106.ena.tsv.gz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f098d614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>taxid</th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>transcript_stable_id</th>\n",
       "      <th>protein_stable_id</th>\n",
       "      <th>primary_accession</th>\n",
       "      <th>secondary_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>CM000685</td>\n",
       "      <td>AAH12389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>chrX</td>\n",
       "      <td>AAH12389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>NC_000023</td>\n",
       "      <td>AAH12389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>ENSP00000482130</td>\n",
       "      <td>CM000685</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>ENSP00000482130</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761259</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000289718</td>\n",
       "      <td>ENST00000696892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761260</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000289718</td>\n",
       "      <td>ENST00000696892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761261</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000289719</td>\n",
       "      <td>ENST00000531488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CM000673</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761262</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000289719</td>\n",
       "      <td>ENST00000531488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761263</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>ENSG00000289719</td>\n",
       "      <td>ENST00000531488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>761264 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             species  taxid   gene_stable_id transcript_stable_id  \\\n",
       "0       Homo_sapiens   9606  ENSG00000000003      ENST00000373020   \n",
       "1       Homo_sapiens   9606  ENSG00000000003      ENST00000373020   \n",
       "2       Homo_sapiens   9606  ENSG00000000003      ENST00000373020   \n",
       "3       Homo_sapiens   9606  ENSG00000000003      ENST00000612152   \n",
       "4       Homo_sapiens   9606  ENSG00000000003      ENST00000612152   \n",
       "...              ...    ...              ...                  ...   \n",
       "761259  Homo_sapiens   9606  ENSG00000289718      ENST00000696892   \n",
       "761260  Homo_sapiens   9606  ENSG00000289718      ENST00000696892   \n",
       "761261  Homo_sapiens   9606  ENSG00000289719      ENST00000531488   \n",
       "761262  Homo_sapiens   9606  ENSG00000289719      ENST00000531488   \n",
       "761263  Homo_sapiens   9606  ENSG00000289719      ENST00000531488   \n",
       "\n",
       "       protein_stable_id primary_accession secondary_accession  \n",
       "0        ENSP00000362111          CM000685            AAH12389  \n",
       "1        ENSP00000362111              chrX            AAH12389  \n",
       "2        ENSP00000362111         NC_000023            AAH12389  \n",
       "3        ENSP00000482130          CM000685                 NaN  \n",
       "4        ENSP00000482130              chrX                 NaN  \n",
       "...                  ...               ...                 ...  \n",
       "761259               NaN             chr11                 NaN  \n",
       "761260               NaN         NC_000011                 NaN  \n",
       "761261               NaN          CM000673                 NaN  \n",
       "761262               NaN             chr11                 NaN  \n",
       "761263               NaN         NC_000011                 NaN  \n",
       "\n",
       "[761264 rows x 7 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "24c05da4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"http://www.informatics.jax.org/downloads/reports/MRK_ENSEMBL.rpt\", sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8cabc50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [\n",
    "    \"MGI Marker Accession ID\",\n",
    "    \"Marker Symbol\",\n",
    "    \"Marker Name\",\n",
    "    \"cM Position\",\n",
    "    \"Chromosome\",\n",
    "    \"Ensembl Accession ID\",\n",
    "    \"Ensembl Transcript ID\",\n",
    "    \"Ensembl Protein ID\",\n",
    "    \"Feature Types\",\n",
    "    \"Genome Coordinate Start\",\n",
    "    \"Genome Coordinate End\",\n",
    "    \"Strand\",\n",
    "    \"BioTypes\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cebbc8b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>cM Position</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Ensembl Accession ID</th>\n",
       "      <th>Ensembl Transcript ID</th>\n",
       "      <th>Ensembl Protein ID</th>\n",
       "      <th>Feature Types</th>\n",
       "      <th>Genome Coordinate Start</th>\n",
       "      <th>Genome Coordinate End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>BioTypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:1919275</td>\n",
       "      <td>1600012P17Rik</td>\n",
       "      <td>RIKEN cDNA 1600012P17 gene</td>\n",
       "      <td>68.49</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000047661</td>\n",
       "      <td>ENSMUST00000062159 ENSMUST00000162474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>158795271</td>\n",
       "      <td>158808032</td>\n",
       "      <td>-</td>\n",
       "      <td>lncRNA|ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:1914753</td>\n",
       "      <td>1700001G17Rik</td>\n",
       "      <td>RIKEN cDNA 1700001G17 gene</td>\n",
       "      <td>12.78</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000103746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>33708905</td>\n",
       "      <td>33709793</td>\n",
       "      <td>+</td>\n",
       "      <td>TEC|ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1916606</td>\n",
       "      <td>1700003I22Rik</td>\n",
       "      <td>RIKEN cDNA 1700003I22 gene</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000100372</td>\n",
       "      <td>ENSMUST00000186048 ENSMUST00000190280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>56058137</td>\n",
       "      <td>56059362</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA|ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1925628</td>\n",
       "      <td>1700006P03Rik</td>\n",
       "      <td>RIKEN cDNA 1700006P03 gene</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000102738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>137253172</td>\n",
       "      <td>137253580</td>\n",
       "      <td>-</td>\n",
       "      <td>TEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1916558</td>\n",
       "      <td>1700007P06Rik</td>\n",
       "      <td>RIKEN cDNA 1700007P06 gene</td>\n",
       "      <td>90.37</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000089730</td>\n",
       "      <td>ENSMUST00000160380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA gene</td>\n",
       "      <td>186857335</td>\n",
       "      <td>186860049</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA|ncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  MGI Marker Accession ID  Marker Symbol                 Marker Name  \\\n",
       "0             MGI:1919275  1600012P17Rik  RIKEN cDNA 1600012P17 gene   \n",
       "1             MGI:1914753  1700001G17Rik  RIKEN cDNA 1700001G17 gene   \n",
       "2             MGI:1916606  1700003I22Rik  RIKEN cDNA 1700003I22 gene   \n",
       "3             MGI:1925628  1700006P03Rik  RIKEN cDNA 1700006P03 gene   \n",
       "4             MGI:1916558  1700007P06Rik  RIKEN cDNA 1700007P06 gene   \n",
       "\n",
       "   cM Position Chromosome Ensembl Accession ID  \\\n",
       "0        68.49          1   ENSMUSG00000047661   \n",
       "1        12.78          1   ENSMUSG00000103746   \n",
       "2        -1.00          1   ENSMUSG00000100372   \n",
       "3        -1.00          1   ENSMUSG00000102738   \n",
       "4        90.37          1   ENSMUSG00000089730   \n",
       "\n",
       "                   Ensembl Transcript ID Ensembl Protein ID Feature Types  \\\n",
       "0  ENSMUST00000062159 ENSMUST00000162474                NaN   lncRNA gene   \n",
       "1                                    NaN                NaN   lncRNA gene   \n",
       "2  ENSMUST00000186048 ENSMUST00000190280                NaN   lncRNA gene   \n",
       "3                                    NaN                NaN   lncRNA gene   \n",
       "4                     ENSMUST00000160380                NaN   lncRNA gene   \n",
       "\n",
       "   Genome Coordinate Start  Genome Coordinate End Strand      BioTypes  \n",
       "0                158795271              158808032      -  lncRNA|ncRNA  \n",
       "1                 33708905               33709793      +     TEC|ncRNA  \n",
       "2                 56058137               56059362      +  lncRNA|ncRNA  \n",
       "3                137253172              137253580      -           TEC  \n",
       "4                186857335              186860049      +  lncRNA|ncRNA  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7ee0b202",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(55340, 13)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "688e43fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    \"http://www.informatics.jax.org/downloads/reports/MGI_EntrezGene.rpt\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    low_memory=False,\n",
    "    dtype={8: pd.Int64Dtype(), 11: pd.Int64Dtype(), 12: pd.Int64Dtype()},\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "e0e6ae2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [\n",
    "    \"MGI Marker Accession ID\",\n",
    "    \"Marker Symbol\",\n",
    "    \"Status\",\n",
    "    \"Marker Name\",\n",
    "    \"cM Position\",\n",
    "    \"Chromosome\",\n",
    "    \"Type\",\n",
    "    \"Secondary Accession IDs\",\n",
    "    \"Entrez Gene ID\",\n",
    "    \"Synonyms\",\n",
    "    \"Feature Types\",\n",
    "    \"Genome Coordinate Start\",\n",
    "    \"Genome Coordinate End\",\n",
    "    \"Strand\",\n",
    "    \"BioTypes\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "2f1d90ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Status</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>cM Position</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Type</th>\n",
       "      <th>Secondary Accession IDs</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Feature Types</th>\n",
       "      <th>Genome Coordinate Start</th>\n",
       "      <th>Genome Coordinate End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>BioTypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>nonagouti</td>\n",
       "      <td>76.83</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50518</td>\n",
       "      <td>agouti|As|agouti signal protein|ASP</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>154792519</td>\n",
       "      <td>154892932</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding|protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>As</td>\n",
       "      <td>W</td>\n",
       "      <td>withdrawn, allele of a</td>\n",
       "      <td>-999.00</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>A&lt;y&gt;</td>\n",
       "      <td>W</td>\n",
       "      <td>withdrawn, allele of a</td>\n",
       "      <td>-999.00</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:87854</td>\n",
       "      <td>Pzp</td>\n",
       "      <td>O</td>\n",
       "      <td>PZP, alpha-2-macroglobulin like</td>\n",
       "      <td>63.02</td>\n",
       "      <td>6</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MGI:2141514</td>\n",
       "      <td>11287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>128460530</td>\n",
       "      <td>128503683</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding|protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:87854</td>\n",
       "      <td>AI893533</td>\n",
       "      <td>W</td>\n",
       "      <td>withdrawn, = A2m</td>\n",
       "      <td>-999.00</td>\n",
       "      <td>6</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  MGI Marker Accession ID Marker Symbol Status  \\\n",
       "0               MGI:87853             a      O   \n",
       "1               MGI:87853            As      W   \n",
       "2               MGI:87853          A<y>      W   \n",
       "3               MGI:87854           Pzp      O   \n",
       "4               MGI:87854      AI893533      W   \n",
       "\n",
       "                       Marker Name  cM Position Chromosome  Type  \\\n",
       "0                        nonagouti        76.83          2  Gene   \n",
       "1           withdrawn, allele of a      -999.00          2  Gene   \n",
       "2           withdrawn, allele of a      -999.00          2  Gene   \n",
       "3  PZP, alpha-2-macroglobulin like        63.02          6  Gene   \n",
       "4                 withdrawn, = A2m      -999.00          6  Gene   \n",
       "\n",
       "  Secondary Accession IDs  Entrez Gene ID  \\\n",
       "0                     NaN           50518   \n",
       "1                     NaN            <NA>   \n",
       "2                     NaN            <NA>   \n",
       "3             MGI:2141514           11287   \n",
       "4                     NaN            <NA>   \n",
       "\n",
       "                              Synonyms        Feature Types  \\\n",
       "0  agouti|As|agouti signal protein|ASP  protein coding gene   \n",
       "1                                  NaN                  NaN   \n",
       "2                                  NaN                  NaN   \n",
       "3                                  NaN  protein coding gene   \n",
       "4                                  NaN                  NaN   \n",
       "\n",
       "   Genome Coordinate Start  Genome Coordinate End Strand  \\\n",
       "0                154792519              154892932      +   \n",
       "1                     <NA>                   <NA>    NaN   \n",
       "2                     <NA>                   <NA>    NaN   \n",
       "3                128460530              128503683      -   \n",
       "4                     <NA>                   <NA>    NaN   \n",
       "\n",
       "                        BioTypes  \n",
       "0  protein_coding|protein-coding  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3  protein_coding|protein-coding  \n",
       "4                            NaN  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "4259ba9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MGI Marker Accession ID     object\n",
       "Marker Symbol               object\n",
       "Status                      object\n",
       "Marker Name                 object\n",
       "cM Position                float64\n",
       "Chromosome                  object\n",
       "Type                        object\n",
       "Secondary Accession IDs     object\n",
       "Entrez Gene ID               Int64\n",
       "Synonyms                    object\n",
       "Feature Types               object\n",
       "Genome Coordinate Start      Int64\n",
       "Genome Coordinate End        Int64\n",
       "Strand                      object\n",
       "BioTypes                    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "7aa71da3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>cM Position</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Type</th>\n",
       "      <th>Secondary Accession IDs</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Feature Types</th>\n",
       "      <th>Genome Coordinate Start</th>\n",
       "      <th>Genome Coordinate End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>BioTypes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Status</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O</th>\n",
       "      <td>674368</td>\n",
       "      <td>674368</td>\n",
       "      <td>674368</td>\n",
       "      <td>674368</td>\n",
       "      <td>674368</td>\n",
       "      <td>674368</td>\n",
       "      <td>12587</td>\n",
       "      <td>71343</td>\n",
       "      <td>224188</td>\n",
       "      <td>674368</td>\n",
       "      <td>633855</td>\n",
       "      <td>633855</td>\n",
       "      <td>597831</td>\n",
       "      <td>162146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>69717</td>\n",
       "      <td>69717</td>\n",
       "      <td>69717</td>\n",
       "      <td>69717</td>\n",
       "      <td>69717</td>\n",
       "      <td>69717</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MGI Marker Accession ID  Marker Symbol  Marker Name  cM Position  \\\n",
       "Status                                                                     \n",
       "O                        674368         674368       674368       674368   \n",
       "W                         69717          69717        69717        69717   \n",
       "\n",
       "        Chromosome    Type  Secondary Accession IDs  Entrez Gene ID  Synonyms  \\\n",
       "Status                                                                          \n",
       "O           674368  674368                    12587           71343    224188   \n",
       "W            69717   69717                        0               0         0   \n",
       "\n",
       "        Feature Types  Genome Coordinate Start  Genome Coordinate End  Strand  \\\n",
       "Status                                                                          \n",
       "O              674368                   633855                 633855  597831   \n",
       "W                   0                        0                      0       0   \n",
       "\n",
       "        BioTypes  \n",
       "Status            \n",
       "O         162146  \n",
       "W              0  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(\"Status\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f8908151",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove withdrawn symbols\n",
    "df = df[df[\"Status\"] == \"O\"].copy()\n",
    "\n",
    "# Remove Cytogenetic Marker as they make Marker Symbol not unique\n",
    "df = df[df[\"Type\"] != \"Cytogenetic Marker\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "6e55f3b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(673102, 15)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "55bca787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI Marker Accession ID</th>\n",
       "      <th>Marker Symbol</th>\n",
       "      <th>Status</th>\n",
       "      <th>Marker Name</th>\n",
       "      <th>cM Position</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Type</th>\n",
       "      <th>Secondary Accession IDs</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Feature Types</th>\n",
       "      <th>Genome Coordinate Start</th>\n",
       "      <th>Genome Coordinate End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>BioTypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:87853</td>\n",
       "      <td>a</td>\n",
       "      <td>O</td>\n",
       "      <td>nonagouti</td>\n",
       "      <td>76.83</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50518</td>\n",
       "      <td>agouti|As|agouti signal protein|ASP</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>154792519</td>\n",
       "      <td>154892932</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding|protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:87854</td>\n",
       "      <td>Pzp</td>\n",
       "      <td>O</td>\n",
       "      <td>PZP, alpha-2-macroglobulin like</td>\n",
       "      <td>63.02</td>\n",
       "      <td>6</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MGI:2141514</td>\n",
       "      <td>11287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>128460530</td>\n",
       "      <td>128503683</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding|protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MGI:87855</td>\n",
       "      <td>Aal</td>\n",
       "      <td>O</td>\n",
       "      <td>active avoidance learning</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>104236</td>\n",
       "      <td>NaN</td>\n",
       "      <td>heritable phenotypic marker</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MGI:87856</td>\n",
       "      <td>Aap</td>\n",
       "      <td>O</td>\n",
       "      <td>active avoidance performance</td>\n",
       "      <td>-999.00</td>\n",
       "      <td>UN</td>\n",
       "      <td>QTL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QTL</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MGI:87859</td>\n",
       "      <td>Abl1</td>\n",
       "      <td>O</td>\n",
       "      <td>c-abl oncogene 1, non-receptor tyrosine kinase</td>\n",
       "      <td>21.86</td>\n",
       "      <td>2</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MGI:2138905|MGI:2443781</td>\n",
       "      <td>11350</td>\n",
       "      <td>c-Abl|E430008G22Rik</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>31578388</td>\n",
       "      <td>31694239</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding|protein-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  MGI Marker Accession ID Marker Symbol Status  \\\n",
       "0               MGI:87853             a      O   \n",
       "3               MGI:87854           Pzp      O   \n",
       "5               MGI:87855           Aal      O   \n",
       "6               MGI:87856           Aap      O   \n",
       "7               MGI:87859          Abl1      O   \n",
       "\n",
       "                                      Marker Name  cM Position Chromosome  \\\n",
       "0                                       nonagouti        76.83          2   \n",
       "3                 PZP, alpha-2-macroglobulin like        63.02          6   \n",
       "5                       active avoidance learning        -1.00          1   \n",
       "6                    active avoidance performance      -999.00         UN   \n",
       "7  c-abl oncogene 1, non-receptor tyrosine kinase        21.86          2   \n",
       "\n",
       "   Type  Secondary Accession IDs  Entrez Gene ID  \\\n",
       "0  Gene                      NaN           50518   \n",
       "3  Gene              MGI:2141514           11287   \n",
       "5  Gene                      NaN          104236   \n",
       "6   QTL                      NaN          107417   \n",
       "7  Gene  MGI:2138905|MGI:2443781           11350   \n",
       "\n",
       "                              Synonyms                Feature Types  \\\n",
       "0  agouti|As|agouti signal protein|ASP          protein coding gene   \n",
       "3                                  NaN          protein coding gene   \n",
       "5                                  NaN  heritable phenotypic marker   \n",
       "6                                  NaN                          QTL   \n",
       "7                  c-Abl|E430008G22Rik          protein coding gene   \n",
       "\n",
       "   Genome Coordinate Start  Genome Coordinate End Strand  \\\n",
       "0                154792519              154892932      +   \n",
       "3                128460530              128503683      -   \n",
       "5                     <NA>                   <NA>    NaN   \n",
       "6                     <NA>                   <NA>    NaN   \n",
       "7                 31578388               31694239      +   \n",
       "\n",
       "                        BioTypes  \n",
       "0  protein_coding|protein-coding  \n",
       "3  protein_coding|protein-coding  \n",
       "5                            NaN  \n",
       "6                            NaN  \n",
       "7  protein_coding|protein-coding  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "cdc50b07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Gene', 'QTL', 'Other Genome Feature', 'Pseudogene',\n",
       "       'Complex/Cluster/Region', 'DNA Segment', 'Transgene',\n",
       "       'BAC/YAC end'], dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "f5771d88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"MGI Marker Accession ID\"].is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "96ee0b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.set_index(\"MGI Marker Accession ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "390ed442",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index().to_feather(\"mgi_complete_set.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f96db6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "ae1fefc8646a06dd2e75004cd934adda7c5727b046986a772e3b44b0ffba9754"
  },
  "kernelspec": {
   "display_name": "Python 3.9.13 ('py39')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "nbproject": {
   "dependency": {
    "nbproject": "0.1a3",
    "pandas": "1.4.2"
   },
   "id": "m41ldPB0kSel",
   "time_init": "2022-06-24T08:37:22.711250+00:00"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
